src/gprs.c

```diff
@@ -3186,21 +3192,32 @@ static void gprs_load_settings(struct ofono_gprs *gprs, const char *imsi)
 static struct pri_context *gprs_context_for_ia(struct ofono_gprs *gprs)
 {
 	GSList *l;
+	struct pri_context *ctx_ia = NULL;
+	struct pri_context *ctx_inet = NULL;	
 
 	for (l = gprs->contexts; l; l = l->next) {
 		struct pri_context *ctx = l->data;
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA)
-			return ctx;
-	}
+		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA) {
+			if (ctx->preferred)
+				return ctx;
 
-	for (l = gprs->contexts; l; l = l->next) {
-		struct pri_context *ctx = l->data;
+			if (ctx_ia == NULL)
+				ctx_ia = ctx;
+		} else if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET &&
```

Comment:
Shouldn't you also be checking for the 'Preferred' flag for Internet context too?  Otherwise, you could end up using a different Internet context than is currently being used by NM.

Also, I'm not sure why you return gprs->contexts->data is neither ctx_ia or ctx_inet is set?  Wouldn't this return a MMS context if it was the only context defined?